lib/go-parser/pclntab.ts

```diff
@@ -0,0 +1,199 @@
+    }
+
+    this.ptrsize = b[7];
+    const uintptr = (b: Buffer): bigint => {
+      if (this.ptrsize === 4) {
+        return this.binary.Uint32(b);
+      }
+      return this.binary.Uint64(b);
+    };
+    const offset = (word: number): bigint => {
+      return uintptr(b.slice(8 + word * this.ptrsize));
+    };
+    const data = (word: number): Buffer => {
+      return b.slice(Number(offset(word))); // TODO: int conversion?
+    };
+
+    switch (this.version) {
+      case pclnVersion.v118:
+        this.nfiletab = Number(offset(1));
+        this.filetab = data(5);
+        this.funcdata = data(7);
+        break;
+      case pclnVersion.v116:
+        this.nfiletab = Number(offset(1));
+        this.filetab = data(4);
+        this.funcdata = data(6);
+        break;
+      case pclnVersion.v12:
+        const nfunctab = Number(this.uintptr(b.slice(8)));
+        this.funcdata = b;
+        const functab = b.slice(8 + this.ptrsize);
+        const functabsize = (nfunctab * 2 + 1) * this.functabFieldSize();
+        const fileoff = this.binary.Uint32(functab.slice(functabsize));
+        this.filetab = b.slice(Number(fileoff));
+        this.nfiletab = Number(this.binary.Uint32(this.filetab));
+        break;
+      default:
+        throw new Error("unreachable");
+    }
+  }
+
+  // go12MapFiles returns a list of files that have been found in the symbol table.
+  // In the original Go implementation, this function takes a map and object, but
+  // we don't need to construct a map and don't need the object.
+  // https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/debug/gosym/pclntab.go;l=669
+  public go12MapFiles(): string[] {
+    this.initFileMap();
+    const files: string[] = [];
+    for (const file of Object.keys(this.fileMap)) {
+      files.push(file);
+    }
+    return files;
+  }
+
+  // uintptr returns the pointer-sized value encoded at b.
+  // The pointer size is dictated by the table being read.
+  private uintptr(b: Buffer): bigint {
+    if (this.ptrsize === 4) {
+      return BigInt(this.binary.Uint32(b));
+    }
+    return this.binary.Uint64(b);
+  }
+
+  // functabFieldSize returns the sivze in bytes of a single functab field.
+  // https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/debug/gosym/pclntab.go;l=377
+  private functabFieldSize(): number {
+    if (this.version >= pclnVersion.v118) {
+      return 4;
+    }
+    return this.ptrsize;
+  }
+
+  // string returns a Go string found at offset.
+  // https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/debug/gosym/pclntab.go;l=372
+  private string(offset: number): string {
+    return this.stringFrom(this.funcdata, offset);
+  }
+
+  // initFileMap initializes the map from file name to file number.
+  // We technically don't need the number, but match Go's implementation 1:1 instead.
+  // https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/debug/gosym/pclntab.go;l=641
+  private initFileMap() {
```

Comment:
can you add return type for this function?

Reply:
I certainly can... I thought not writing anything means we don't return anything... 🙃